File: src/test/src/finalv1.py
#!/usr/bin/env python3
import cv2
import torch
import rospy
from std_msgs.msg import String
import pyrealsense2 as rs2
import statistics as st

from test.realsense_camera import RealsenseCamera
rs=RealsenseCamera()
point = [0,0]
xA=0
xB=0
yA=0
yB=0
distance =0
dis=0
model = torch.hub.load('ultralytics/yolov5', 'custom', path='/home/light/re_ws/src/test/src/yolov5-best.pt', force_reload=True)
def callback(data):
    rospy.loginfo(rospy.get_caller_id() + "I heard %s", data.data)


def talker():
    pub = rospy.Publisher('chatter', String, queue_size=10)
    rospy.init_node('talker', anonymous=True)
    rate = rospy.Rate(10) # 10hz
    while not rospy.is_shutdown():
        dis=processing()
       
        
        hello_str = dis % rospy.get_time()
        rospy.loginfo(hello_str/1000)
        pub.publish(str(hello_str/1000))
        
        rate.sleep()
def processing():
    global distance,dis,classs,object,distance_bc
    while True:
        try:
            ret,color,depth=rs.get_frame_stream()
            if ret == True:
                 print("no frame recevied")
            results = model(color)
            for box in results.xyxy[0]:     
                            xB = int(box[2])
                            xA = int(box[0])
                            yB = int(box[3])
                            yA = int(box[1])
                            classs= int(box[5])
                            center_x= (xA+xB)/2
                            center_y=(yA+yB)/2

                            # print(center_x)
                            # print(center_y)
                            point[0]=int(center_x)
                            point[1]=int(center_y)
                            distance =depth[point[1],point[0]]
                            distance_bc=depth[yB,xB]
  
            if classs == 0 :
                  object = 'CARPET'
            if classs == 1 :
                  object = 'GLASS'
            if classs == 2 :
                  object = 'RAMP' 
            if classs == 3 :
                  object = 'WIRES'                                              
            print("-------------------------------------------->distance  BC",distance_bc)
            cv2.putText(color, "{}cm".format(distance/10), ( point[0],  point[1] - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 2)
            cv2.putText(color, object, ( point[0]-50,  point[1] - 50), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 2)
            cv2.circle(color ,point,4,(0,0,255)) 
            cv2.rectangle(color, (xA, yA), (xB, yB), (0, 255, 0), 2)
            
            cv2.imshow("YOLO Detection", color)
           
            if cv2.waitKey(1) == ord('q'):
                 rs.release()
                 cv2.destroyAllWindowstalker
                 break
            
        except:
            print("out of range")
            pass
        return distance
        
    
if __name__=='__main__':
       talker()
    #    listener()
    # processing()
  

File: src/src/final_v1.cpp
// This program uses OpenCV, PyTorch, and ROS libraries
#include <opencv2/opencv.hpp>
#include <torch/torch.h>
#include <ros/ros.h>
#include <std_msgs/String.h>
#include <realsense2/rs.hpp>
#include <statistics.h>

#include "test/realsense_camera.h"
RealsenseCamera rs;
int point[2] = {0, 0};
int xA = 0;
int xB = 0;
int yA = 0;
int yB = 0;
int distance = 0;
auto model = torch::hub::load("ultralytics/yolov5", "custom", "/home/light/re_ws/src/test/src/yolov5-best.pt", true);

void callback(const std_msgs::String::ConstPtr& data)
{
    ROS_INFO("I heard: [%s]", data->data.c_str());
}

void talker()
{
    ros::Publisher pub = n.advertise<std_msgs::String>("chatter", 10);
    ros::NodeHandle n;
    ros::Rate rate(10); // 10hz
    while (ros::ok())
    {
        int dis = processing();
        std::stringstream ss;
        ss << dis << " at time " << ros::Time::now();
        std_msgs::String msg;
        msg.data = ss.str();
        ROS_INFO("%s", msg.data.c_str());
        pub.publish(msg);
        ros::spinOnce();
        rate.sleep();
    }
}

int processing()
{
    global distance;
    while (true)
    {
        try
        {
            bool ret;
            cv::Mat color, depth;
            std::tie(ret, color, depth) = rs.get_frame_stream();
            if (ret == true)
            {
                std::cout << "no frame received" << std::endl;
            }
            auto results = model.forward(color);
            for (auto box : results.xyxy[0])
            {
                xB = int(box[2]);
                xA = int(box[0]);
                yB = int(box[3]);
                yA = int(box[1]);
                int center_x = (xA + xB) / 2;
                int center_y = (yA + yB) / 2;
                // std::cout << center_x << std::endl;
                // std::cout << center_y << std::endl;
                point[0] = int(center_x);
                point[1] = int(center_y);
            }

            distance = depth.at<int>(point[0], point[1]);
            // std::cout << "beforemsg->info.width median" << distance << std::endl;
            // std::cout << "type---------->" << typeid(depth).name() << std::endl;
            // int dis=    st.median(distance);
            // std::cout << "after median" << dis << std::endl;
            std::cout << results.pandas().xyxy[0] << std::endl;
            // std::cout << "new distance ---------->" << rs.get_distance(point[0],point[1]) << std::endl;
            cv::putText(color, std::to_string(distance / 10) + "cm", cv::Point(point[0], point[1] - 10), cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 255, 0), 2);
            cv::circle(color, cv::Point(point[0], point[1]), 4, cv::Scalar(0, 0, 255));
            cv2.rectangle(color, cv2.Point(xA, yA), cv2.Point(xB, yB), cv2.Scalar(0, 255, 0), 2);

            cv2.imshow("YOLO Detection", color);
            // cv2.imshow("YOLO Detectin", depth);
            if (cv2.waitKey(1) == 'q')
            {
                rs.release();
                cv2.destroyAllWindows();
                break;
            }
        }
        catch (...)
        {
            std::cout << "out of range" << std::endl;
        }
        return distance;
    }
}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "talker");
    talker();
    // listener();
    // processing();
}

